aws-java-sdk-sts/src/main/java/com/amazonaws/auth/STSSessionCredentialsProvider.java

```diff
@@ -119,36 +160,119 @@ public STSSessionCredentialsProvider(AWSCredentialsProvider longLivedCredentials
      */
     public void setSTSClientEndpoint(String endpoint) {
         securityTokenService.setEndpoint(endpoint);
-        sessionCredentials = null;
+        sessionCredentials.set(null);
     }
 
     @Override
     public AWSSessionCredentials getCredentials() {
         if (needsNewSession()) startSession();
 
-        return sessionCredentials;
+        return sessionCredentials.get().sessionCredentials;
     }
 
     @Override
     public void refresh() {
-        startSession();
+        asyncNewSession();
     }
+    
+    /**
+     * Atomically create a new session in a lock-free manner. If an exception
+     * is not thrown, it means exactly one thread successfully set the new
+     * current session, but not necessarily the calling thread
+     */
+    private void casNewSession() {
+        SessionCredentialsHolder currentSession = sessionCredentials.get();
+
+        if (null == currentSession || expiring(currentSession.sessionCredentialsExpiration)) {
+            GetSessionTokenResult sessionTokenResult = securityTokenService
+                    .getSessionToken(new GetSessionTokenRequest()
+                            .withDurationSeconds(DEFAULT_DURATION_SECONDS));
 
+            Credentials stsCredentials = sessionTokenResult.getCredentials();
+
+            SessionCredentialsHolder newSession = new SessionCredentialsHolder()
+                    .withSessionCredentials(new BasicSessionCredentials(
+                                    stsCredentials.getAccessKeyId(),
+                                    stsCredentials.getSecretAccessKey(),
+                                    stsCredentials.getSessionToken()))
+                    .withSessionCredentialsExpiration(stsCredentials.getExpiration());
+            
+            sessionCredentials.compareAndSet(currentSession, newSession);
+        }
+    }
+    
+    /**
+     * Used when there is no valid session to return. Only called
+     * for initial or expired session. Callers are blocked until
+     * exactly one new session is created or an exception is thrown.
+     */
+    private synchronized void serialzedNewSession() {
+        while (expired()) {
```

Comment:
I don't think we should loop indefinitely here. If there is a problem retrieving credentials from STS any thread call getCredentials could hang forever. How about instead let's try and acquire the refreshing lock and give up after a certain time (say five seconds). I think this will cover 99.99% of cases where the first request succeeds in a timely fashion but doesn't hang forever in the .01% chance that the there is a problem. Even with this approach there is a possibility that we will hammer STS with calls once the timeout expires but I think that's the exceptional case and not the norm.